envoy/matcher/matcher.h

```diff
@@ -25,7 +25,15 @@ class ServerFactoryContext;
 
 namespace Matcher {
 
-using MatchingDataType = absl::variant<absl::monostate, std::string>;
+// Abstract interface for custom matching data.
+// Overrides this interface to provide custom matcher specific implementation.
+class CustomMatchData {
+public:
+  virtual ~CustomMatchData() = default;
+};
+
+using MatchingDataType =
+    absl::variant<absl::monostate, std::string, std::shared_ptr<CustomMatchData>>;
```

Comment:
Can we avoid shared_ptr here? Is there a reason why matching data is copied?

Reply:
It would be faster and easier code with unique_ptr. I'm not sure if we can avoid shared - would need to track where the ownership is split.

Reply:
yea it would be slightly faster, but i think current code is more flexible. As the interface, if user don't want to copy, they can std::move. I actually used `unique_ptr` throughout my CEL development until I ran the test using test_utility.h.
If user want to the pattern like what test_utility does, interface just allow it.